### PR TITLE
Add bootstrap check docs

### PR DESCRIPTION
**Pull request by jasontedor**
This commit adds documentation for the bootstrap checks and provides
either links or inline guidance for setting the necessary settings to
pass the bootstrap checks.

### CODE REVIEW

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 1]
```diff
@@ -0,0 +1,144 @@
+[[bootstrap-checks]] == Bootstrap Checks
+
+Collectively, we have a lot of experience with users suffering
+unexpected issues because they have not configured
+<<important-setting,important settings>>. In previous versions of
+Elasticsearch, some of these settings were logged as
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: In previous versions of Elasticsearch, *forgetting to set* some of these settings caused Elasticsearch to log warnings.

Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: Or something.

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+<<[[bootstrap.mlockall]] is enabled, the JVM will lock the initial size
+of the heap on startup. If the initial heap size is not equal to the
+maximum heap size, after a resize it will not be the case that all of
+the JVM heap is locked in memory. To pass the heap size check, you must
+configure the [[heap-size,heap size]].
+
+=== File descriptor check
+
+File descriptors are a Unix construct for tracking open "files".
+"Files" though could be a physical file, a virtual file (e.g.,
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I think the phrase "[everything is a file](https://en.wikipedia.org/wiki/Everything_is_a_file)" is fairly common and probably deserves repeating here somewhere.

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 1 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I pushed 430ca4b88a056487dddbdf2cd8accf8b14a3263b.

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+``/proc/loadavg`), or network sockets. Elasticsearch requires lots file
+descriptors (e.g., every shard is composed of multiple segments and
+other files, plus connections to other nodes, etc.). This bootstrap
+check is enforced on OS X and Linux. To pass the file descriptor check,
+you might have to configure [[file-descriptors,file descriptors]].
+
+=== Memory lock check
+
+A terrible thing can happen if parts of the JVM heap are swapped out to
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I might reword "When the JVM does a major garbage collection it touches every page of the heap. If any of those pages are swapped out to disk they will have to be swapped back in to memory. That causes lots of disk activity that Elasticsearch would much rather be used to service request...."

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+configure <<mlockall,`mlockall`>>.
+
+=== Minimum master nodes check
+
+Elasticsearch uses a single master for managing cluster state but
+enables there to be multiple master-eligible nodes for
+high-availability. In the case of a partition, master-eligible nodes on
+each side of the partition might be elected as the acting master without
+knowing that there is a master on the side of the partition. This can
+lead to divergent cluster states potentially leading to data loss when
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: Do diverging cluster states ever *not* lead to data loss?

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 2 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: Good suggestion. I pushed a8b33185f3c6b7418b3f90980988cd6ade3f8e5a.

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+master-eligible nodes, it is not possible for the cluster to suffer from
+split brain because during a network partition there can be at most one
+side of the partition that contains a quorum of master nodes. The
+minimum master nodes check enforces that you've set
+<<minimum_master_nodes>>. To pass the minimum master nodes check, you
+must configure <<minimum_master_nodes>>.
+
+NOTE: The minimum master nodes check does not enforce that you've
+configured <<minimum_master_nodes>> correctly, only that you have
+configured. Elasticsearch does log a warning message if it detects that
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: **it** configured

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+create a lot of threads. The maximum number of threads check ensures
+that the Elasticsearch process has the rights to create enough threads
+under normal use. This check is enforced only on Linux. If you are on
+Linux, to pass the maximum number of threads check, you must configure
+your system to allow the Elasticsearch process the ability to create at
+least 2048 threads. This can be done via `/etc/security/limits.conf`
+using the `nproc` setting (note that you might have to increase the
+limits for the `root` user too).
+
+[[max-size-virtual-memory-check]] === Maximum size virtual memory check
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I think the heading should be on a new line?

File: docs/reference/setup/bootstrap-checks.asciidoc [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+Continuing from the previous <<max-size-virtual-memory-check,point>>, to
+use `mmap` effectively, Elasticsearch also requires the ability to
+create many memory-mapped areas. The maximum map count check checks that
+the kernel allows a process to have at least 262,144 memory-mapped areas
+and is enforced on Linux only. To pass the maximum map count check, you
+must configure `vm.max_map_count` via `sysctl` to be at least `262144`.
+
+=== Client JVM check
+
+There are two different JVMs provided by OpenJDK: the client JVM and the
```
Comment (nik9000) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I think --client is also an Oracle thing, right? Not just OpenJDK?

Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I meant to say the OpenJDK-derived JDKs. Will clarify.

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 6 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I pushed 76bbe41398fa8750f43b51df8b114c061d667f25.

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 5 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I pushed 334aca9b357549181c1ed79fe938cfcf7de0128c.

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 3 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: Thanks for the suggestion. I pushed 9bf872ed98e82f121ae590ab966a79faccb0955d.

File: docs/reference/setup/bootstrap-checks.asciidoc (hunk 7 above)
Comment (jasontedor) [inline docs/reference/setup/bootstrap-checks.asciidoc]: I pushed 04108eb7d1fc520cf40cd60392655b4e06583549.

**Comment by nik9000** [comment]
LGTM

**Comment by clintongormley** [comment]
@jasontedor this PR doesn't mention the escape hatch added in https://github.com/elastic/elasticsearch/pull/18088

**Pull request merged**